README.md
<h1 align="center">
  <img alt="podcastr" title="podcastr" src=".github/Logo.png" />
</h1>

<div align="center">

![Website](https://img.shields.io/website?style=flat&up_message=online&labelColor=000000&url=https%3A%2F%2Fletmeask-9b20f.web.app%2F)
![Badge](https://img.shields.io/github/license/PedroHenry-Santos/Letmeask?style=flat&label=LICENSE&color=FFC300&labelColor=000000)
![GitHub release (latest by date)](https://img.shields.io/github/v/release/PedroHenry-Santos/letmeask?style=flat&labelColor=000000&color=FFC300)
![GitHub last commit](https://img.shields.io/github/last-commit/PedroHenry-Santos/letmeask?color=FFC300&style=flat&labelColor=000000)
![GitHub repo size](https://img.shields.io/github/repo-size/Pedrohenry-Santos/letmeask?color=FFC300&labelColor=000000)


![Badge](https://img.shields.io/badge/TYPESCRIPT-v4.2.4-blue?style=flat&color=298BF9&labelColor=000000&logo=TypeScript)
![Badge](https://img.shields.io/badge/REACT-v17.0.2-blue?style=flat&color=298BF9&labelColor=000000&logo=react)
![Badge](https://img.shields.io/badge/Firebase-v8.6.8-blue?style=flat&logo=Firebase&color=298BF9&labelColor=000000)
![Badge](https://img.shields.io/badge/Styled--components-v5.3.0-blue?style=flat&logo=styled-components&labelColor=000000&color=298BF9)
![Badge](https://img.shields.io/badge/Framer--motion-v4.1.17-blue?style=flat&logo=framer&labelColor=000000&color=298BF9)

</div>

<div align="center">
  <img src=".github/demo.gif" with="800" height="400"/>
</div>

# :page_with_curl: Índice

* [Sobre o projeto](#interrobang-sobre-o-projeto)
  - [Tecnologias utilizadas](#-tecnologias-utilizadas)
* [Executando a aplicação](#rocket-executando-a-aplicação)
  - [Executando a aplicação](#pencil2-Executando-a-aplicação)
* [Para contribuir](#mailbox-para-contribuir)
  - [Histórico de atualizações](#clock10-histórico-de-atualizações)

# :interrobang: Sobre o projeto

Projeto realizado durante a NLW#6 realizado pela Rocketseat. Foi desenvolvido um app para onde é possível realizar interação com usuários através de perguntas.
## 🧑‍💻 Tecnologias utilizadas

- [React](https://reactjs.org)
- [TypeScript](https://www.typescriptlang.org/)
- [Styled-components](https://styled-components.com/)
- [Firebase](https://firebase.google.com/)
# :rocket: Executando a aplicação

Primeiramente clone o repositório para sua maquina ou faça o 'download' em formato ZIP e descompacte. Para realizar a clonagem do repositório utilize o comando:

    git clone git@github.com:PedroHenry-Santos/Letmeask.git

Após basta instalar as dependência utilizando o gerenciador de pacote de sua preferência. Conforme abaixo:

    yarn install

E utilizar o comando:

    yarn start

Agora você pode acessar [`localhost:3000`](http://localhost:3000) do seu navegador.

Lembrando que será necessário criar uma conta no [Firebase](https://firebase.google.com/) e um projeto para disponibilizar um Realtime Database.
# :mailbox: Para contribuir

Caso queira ajudar a implementar novas funcionalidade nessa aplicação, poderá seguir o **roteiro de funcionalidades** a seguir. Ou também poderá propor sua própria funcionalidade.<br/><br/>
Para isso basta seguir o [roteiro](https://github.com/PedroHenry-Santos/Letmeask/blob/main/CONTRIBUTING.md) em CONTRIBUTING.

## :heavy_check_mark: Roteiro de funcionalidade da aplicação

Para um acompanhamento mais atualizado verifique a aba [Projeto](https://github.com/PedroHenry-Santos/Letmeask/projects/1).

- [x] Autenticação através de conta Google utilizando o Firebase. (:memo: Realizado durante o evento)
- [x] Criar sala de perguntas individuais. (:memo: Realizado durante o evento)
- [x] Adicionar novas perguntas na sala. (:memo: Realizado durante o evento)
- [x] Pagina do administrador das perguntas. (:memo: Realizado durante o eventos)
- [x] Like nas perguntas. (:memo: Realizado durante o evento)
- [x] Remover perguntas. (:memo: Realizado durante o evento)
- [x] Encerrar sala. (:memo: Realizado durante o evento)
- [x] Impedir que entre em uma sala fechada com código. (:memo: Realizado durante o evento)
- [x] Função de dar destaque a uma pergunta. (:memo: Realizado durante o evento)
- [x] Função de marcar pergunta respondida. (:memo: Realizado durante o evento)
- [x] Toast para notificação. (:rocket: Milhas extras...)
- [x] Modal de confirmação para exclusão e encerramento da chamada. (:rocket: Milhas extras...)
- [x] Temas de exibição (Light, Dark). (:rocket: Milhas extras...)
- [x] Animações na tela inicial e em todos os botões da aplicação com o [framer-motion](https://www.framer.com/motion/). (:rocket: Milhas extras...)
- [x] Mostrar se o usuário já esta autenticado na tela inicial e de criação de salas. (:rocket: Milhas extras...)
- [x] Funcionalidade de fechamento automático da sala para todos os usuários quando a sala for encerrada. (:rocket: Milhas extras...)
- [x] Direcionamento automático do administrador para a sala correspondente quando colocar o código na home. (:rocket: Milhas extras...)
- [x] As perguntas mais bem votadas são colocadas em primeiro lugar. (:rocket: Milhas extras...)
- [x] A pergunta respondida vai para à ultima posição. (:rocket: Milhas extras...)
- [x] A pergunta destacada vai para à primeira posição. (:rocket: Milhas extras...)
- [x] Funcionalidade para filtrar as 20 primeiras questões com firebase. (:rocket: Milhas extras...)
- [x] Responsividade. (:rocket: Milhas extras...)

## :clock10: Histórico de atualizações

Para saber sobre as [versões](https://github.com/PedroHenry-Santos/Letmeask/releases) deste código consulte o CHANGELOG detalhado.

# :lock: Licença

[MIT License](https://github.com/PedroHenry-Santos/Letmeask/blob/main/LICENCE.md) ® [Pedro Santos](https://github.com/PedroHenry-Santos)


CHANGELOG.md
# [1.5.0](https://github.com/PedroHenry-Santos/Letmeask/compare/v1.4.0...v1.5.0) (2021-06-30)


### Bug Fixes

* **room:** fix production bug, where it does not close the room in some cases ([93aae32](https://github.com/PedroHenry-Santos/Letmeask/commit/93aae32a4d9158f8100a7651e97da9f7a42dd7f8))
* **room/adminroom:** fix bug in the styling of room components ([535a9be](https://github.com/PedroHenry-Santos/Letmeask/commit/535a9be300fecc98fadccf15fbc795bb645f7674))


### Features

* **responsivo:** add responsive to all pages and components ([6b2dafd](https://github.com/PedroHenry-Santos/Letmeask/commit/6b2dafd78a09919d5e71768a7adf910c7875609d))
* **room/adminroom:** add answered questions as last positions ([fec833d](https://github.com/PedroHenry-Santos/Letmeask/commit/fec833d703f3067e4a96330d998559c97bbad6cf))
* **room/adminroom:** add limit of questions to be displayed on screen ([04c48ae](https://github.com/PedroHenry-Santos/Letmeask/commit/04c48ae8cf1a4c3e99788c93a282c51e7b780bcf))
* **room/adminroom:** add sorting of questions by likes in real time ([d8cfc95](https://github.com/PedroHenry-Santos/Letmeask/commit/d8cfc95052afab8a92ad6a06266205bcfb7f84a2))
* **room/adminroom:** add the highlighted questions as the first ([98994cc](https://github.com/PedroHenry-Santos/Letmeask/commit/98994ccd8bd68a8ed4bb57f87126c6a8131bccee))

# [1.4.0](https://github.com/PedroHenry-Santos/Letmeask/compare/v1.3.0...v1.4.0) (2021-06-28)


### Features

* **closepage:** add redirect page when room is closed and refactor images ([034cd0b](https://github.com/PedroHenry-Santos/Letmeask/commit/034cd0bc755df22994f4f0b94274bdc6da321a55))
* **home:** add automatic redirection of users according to their permission ([35a0374](https://github.com/PedroHenry-Santos/Letmeask/commit/35a0374555d837195d10282fd6f6875ea00ba47e))
* **room/adminroom:** feat to check if the room is active and close the room to all users realtime ([0877c54](https://github.com/PedroHenry-Santos/Letmeask/commit/0877c547a5c3fed5f0c8e5f56431e51b5cf6c842))

# [1.3.0](https://github.com/PedroHenry-Santos/Letmeask/compare/v1.2.0...v1.3.0) (2021-06-28)


### Features

* **home/newroom:** add component shows the logged in user, changing the interface to user ([7bdbdd1](https://github.com/PedroHenry-Santos/Letmeask/commit/7bdbdd14185ff6338be79749dba48ffdca58f1b4))
* **motion:** add animation to components with framer-motion for more interactivity ([04d77f4](https://github.com/PedroHenry-Santos/Letmeask/commit/04d77f4d2496c5320e2e67eafe590e4bb810ff83))

# [1.2.0](https://github.com/PedroHenry-Santos/Letmeask/compare/v1.1.0...v1.2.0) (2021-06-27)


### Bug Fixes

* **room:** fix login bug in question room ([d273b71](https://github.com/PedroHenry-Santos/Letmeask/commit/d273b7126dcdc4a2398410ebb18459e7eb086fc5))
* **themes:** fix theme bug in modal ([e9da256](https://github.com/PedroHenry-Santos/Letmeask/commit/e9da256960e99fc5a6867e61fd7499c6bd3f8fc6))


### Features

* **newroom/room:** add create new room, create question room page, create and list new question ([49c9988](https://github.com/PedroHenry-Santos/Letmeask/commit/49c99883bf6f89efbfdaf0eb88d3aefec08ef1c9))
* **room/adminroom:** add feature to highlight and mark answered questions ([8194b58](https://github.com/PedroHenry-Santos/Letmeask/commit/8194b588dca0a7fa607d2d87fbacd6a79ea2e361))
* **room/adminroom:** add presentation image in question rooms ([fa1a8c0](https://github.com/PedroHenry-Santos/Letmeask/commit/fa1a8c0d377c4bc67ead3c9e5db1199684fbd8d8))
* **room/adminroom:** add question component, share info, admin page, remove questions and end room ([05e7ac5](https://github.com/PedroHenry-Santos/Letmeask/commit/05e7ac52ed1c3ffe926a6c58c04acf0b2fde4a64))
* **themes:** add dark and light theme functionality and Scss refactoring for stylized-components ([f604313](https://github.com/PedroHenry-Santos/Letmeask/commit/f60431391db905df28b4c50e0ad8a5b02c804902))

# [1.1.0](https://github.com/PedroHenry-Santos/Letmeask/compare/v1.0.0...v1.1.0) (2021-06-24)


### Features

* **newroom/room:** add create new room, create question room page, create and list new question ([20b529e](https://github.com/PedroHenry-Santos/Letmeask/commit/20b529e0c4157ba7caa85e2fc50a0d99a53ea5bc))

# 1.0.0 (2021-06-23)


### Features

* **home/newroom:** add home pages, Firebase authentication and context to share information ([63e9d4c](https://github.com/PedroHenry-Santos/Letmeask/commit/63e9d4c9de9207078153adbfc62a8c84d780ee10))


src/assets/styles/room.styles.ts
import styled from 'styled-components';

export const RoomStyle = styled.div`
    display: grid;
    grid-template-columns: 1fr;
    grid-template-rows: 5.2rem 1fr;
    height: 100vh;
    width: 100vw;
    overflow: hidden;
`;

export const Header = styled.header`
    width: 100vw;
    height: 5.2rem;
    padding: 0 1.6rem;
    border-bottom: 1px solid
        ${props => props.theme.colors.backgroundAnsweredQuestion};
    background: ${props => props.theme.colors.headerBackground};
    justify-content: center;
    align-items: center;

    .content {
        width: 100%;
        height: 100%;
        max-width: 74.7rem;
        margin: 0 auto;
        display: flex;
        justify-content: space-between;
        align-items: center;

        a {
            display: flex;

            > img {
                justify-self: center;
                align-self: center;
                max-height: 3rem;
            }
        }

        .theme-mode {
            margin: auto 0;
            display: flex;
            flex-direction: column;
            gap: 0.4rem;

            span {
                font-size: 0.9rem;
                font-weight: 500;
            }
        }

        > button {
            > span {
                display: none;

                @media (min-width: 700px) {
                    display: initial;
                }
            }
        }

        > div {
            display: flex;
            gap: 1.1rem;

            button {
                height: 2.7rem;
                background: transparent;

                &:hover {
                    color: ${props => props.theme.colors.whiteText};
                }
            }
        }
    }
`;

export const MainAdminRoom = styled.main`
    display: grid;
    grid-template-columns: 1fr;
    grid-template-rows: 22% 78%;
    margin: 0 auto;
    padding: 0 2rem;
    height: 100%;
    width: 100%;
    max-width: 53.4rem;

    @media (min-width: 500px) {
        grid-template-rows: 24% 76%;
    }

    @media (min-width: 600) {
        grid-template-rows: 26% 74%;
    }

    @media (min-width: 800) {
        grid-template-rows: 28% 72%;
    }
`;

export const MainRoom = styled.main`
    display: grid;
    grid-template-columns: 1fr;
    grid-template-rows: 27% 26% 47%;
    margin: 0 auto;
    padding: 0 2rem;
    height: 100%;
    width: 100%;
    max-width: 53.4rem;

    @media (min-width: 400px) {
        grid-template-rows: 22% 26% 52%;
    }

    @media (min-height: 800px) {
        grid-template-rows: 20% 24% 56%;
    }

    @media (min-width: 1200px) {
        grid-template-rows: 20% 30% 50%;
    }
`;

export const RoomTitle = styled.section`
    display: flex;
    align-items: flex-end;
    margin-bottom: 1.7rem;

    @media (min-width: 700px) {
        justify-content: space-between;
    }

    .info-questions {
        display: flex;
        flex-direction: column;
        justify-content: flex-end;
        align-items: flex-start;
        height: 100%;

        h1 {
            display: flex;
            margin-top: auto;
            max-width: calc(100vw - 5rem);
            font-family: 'Poppins', sans-serif;
            font-size: 1.2rem;
            color: ${props => props.theme.colors.mainText};
            overflow-x: hidden;
            text-overflow: ellipsis;

            @media (min-width: 600px) {
                font-size: 1.6rem;
            }
        }

        @media (min-width: 600px) {
            display: flex;
            align-items: flex-end;
            flex-direction: initial;

            h1 {
                max-width: 30rem;
                white-space: normal;
                word-wrap: break-word;
                word-break: keep-all;
                text-align: justify;
            }
        }

        span {
            background: ${props => props.theme.colors.secondary};
            border-radius: 9999px;
            padding: 0.6rem 1.1rem;
            color: ${props => props.theme.colors.whiteText};
            font-weight: 500;
            font-size: 0.9rem;
            margin-top: 0.6rem;

            @media (min-width: 600px) {
                margin-left: 1.1rem;
            }
        }
    }

    .theme-mode {
        margin: auto 0;
        display: flex;
        flex-direction: column;
        gap: 0.4rem;

        span {
            font-size: 0.9rem;
            font-weight: 500;
        }
    }
`;

export const Form = styled.section`
    display: flex;
    flex-direction: column;

    textarea {
        width: 100%;
        min-width: 8.7rem;
        height: 6rem;
        resize: none;
        border: 0;
        padding: 1.1rem 1.1rem;
        border-radius: 0.6rem;
        background: ${props => props.theme.colors.interactionBackground};
        box-shadow: 0 2px 0.8rem rgba(0, 0, 0, 0.04);
        color: ${props => props.theme.colors.mainText};

        @media (min-width: 400px) {
            height: 8.7rem;
        }

        @media (min-height: 800px) {
            height: 8.7rem;
        }

        @media (min-height: 1200px) {
            height: 14rem;
            font-size: 2rem;
        }

        &:focus {
            outline: none !important;
            border-color: ${props => props.theme.colors.primary};
            box-shadow: 0 0 0.33rem ${props => props.theme.colors.primary};
        }

        &::-webkit-scrollbar {
            width: 0.5rem;
            background-color: ${props =>
                props.theme.colors.interactionBackground};
        }

        &::-webkit-scrollbar-thumb {
            border-radius: 9999px;
            width: 0.4rem !important;
            background: red;
            background-color: ${props =>
                props.theme.colors.backgroundAnsweredQuestion};
        }
    }

    .form-footer {
        display: flex;
        justify-content: space-between;
        align-items: center;
        margin-top: 1rem;

        .user-info {
            display: flex;
            align-items: center;

            img {
                width: 2.2rem;
                height: 2.2rem;
                border-radius: 50%;
            }

            span {
                margin-left: 0.6rem;
                color: ${props => props.theme.colors.mainText};
                font-weight: 500;
                font-size: 0.9rem;
            }
        }

        > span {
            font-size: 0.9rem;
            color: ${props => props.theme.colors.secondaryText};
            font-weight: 500;

            button {
                background: transparent;
                border: 0;
                color: ${props => props.theme.colors.primary};
                text-decoration: underline;
                font-size: 0.9rem;
                font-weight: 500;
                cursor: pointer;
            }
        }
    }
`;

export const QuestionList = styled.section`
    padding: 0 0.5rem;
    border-radius: 0.8rem;
    overflow: auto;

    &::-webkit-scrollbar {
        width: 0.5rem;
        background-color: ${props => props.theme.colors.background};
    }

    &::-webkit-scrollbar-thumb {
        border-radius: 9999px;
        width: 0.4rem !important;
        background: red;
        background-color: ${props =>
            props.theme.colors.backgroundAnsweredQuestion};
    }
`;

export const EmptyList = styled.section`
    padding: 5rem 0 0 0;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;

    img {
        width: 12rem;
        height: 12rem;
    }

    h3 {
        padding: 1rem 0;
        font-size: 1.2rem;
        font-weight: 600;
        font-family: 'Poppins', sans-serif;
        color: ${props => props.theme.colors.mainText};
    }

    p {
        width: 20rem;
        text-align: center;
        font-size: 0.9rem;
        line-height: 1.4rem;
        color: ${props => props.theme.colors.secondaryText};
    }
`;

export const ModalStyle = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    padding: 2rem 3.33rem;

    > svg {
        height: 3.2rem;
        width: 3.2rem;

        path {
            stroke: ${props => props.theme.colors.alertButton};
        }
    }

    h2 {
        padding-top: 1rem;
        font-size: 1.6rem;
        font-family: 'Poppins', sans-serif;
        font-weight: 700;
        line-height: 2.3rem;
        color: ${props => props.theme.colors.mainText};
    }

    p {
        font-family: 'Roboto', sans-serif;
        font-size: 1.1rem;
        line-height: 1.7rem;
        text-align: center;
        color: ${props => props.theme.colors.secondaryText};
        padding: 0.7rem;
    }

    > div {
        gap: 10px;
        display: flex;

        button {
            height: 3.33rem;
            border-radius: 0.6rem;
            font-weight: 500;
            padding: 0 2.2rem;
            display: flex;
            justify-content: center;
            align-items: center;
            cursor: pointer;
            border: 0;
            transition: filter 0.2s;

            &:hover {
                filter: brightness(0.9);
            }
        }

        .cancel {
            background: ${props => props.theme.colors.disabledButton};
            color: ${props => props.theme.colors.secondaryText};

            &:hover {
                background: ${props => props.theme.colors.disabledButtonHover};
            }
        }

        .confirm {
            background: ${props => props.theme.colors.alertButton};
            color: ${props => props.theme.colors.whiteText};

            &:hover {
                background: ${props => props.theme.colors.alertButtonHover};
            }
        }
    }
`;
